File: evidence/v1_1/ai_eval/subjects/counted/r1/sieve/cpp/main.cpp
#include <algorithm>
#include <cmath>
#include <cstdint>
#include <iostream>
#include <thread>
#include <vector>

int main() {
    long long n;
    if (!(std::cin >> n)) {
        return 1;
    }
    if (n < 0) n = 0;

    if (n < 2) {
        std::cout << "count 0\n";
        std::cout << "sum 0\n";
        std::cout << "last none\n";
        std::cout << "gap none\n";
        return 0;
    }

    // is_composite[i] true means i is not prime. Index range [0, n].
    std::vector<uint8_t> is_composite(static_cast<size_t>(n) + 1, 0);
    is_composite[0] = 1;
    is_composite[1] = 1;

    // Base primes up to sqrt(n), computed single-threaded (small work).
    long long limit = static_cast<long long>(std::sqrt(static_cast<double>(n))) + 1;
    while (limit * limit > n) --limit;
    while ((limit + 1) * (limit + 1) <= n) ++limit;

    std::vector<uint8_t> base_composite(static_cast<size_t>(limit) + 1, 0);
    if (limit >= 0) {
        if (limit >= 0) base_composite[0] = 1;
        if (limit >= 1) base_composite[1] = 1;
    }
    for (long long p = 2; p * p <= limit; ++p) {
        if (!base_composite[p]) {
            for (long long m = p * p; m <= limit; m += p) {
                base_composite[m] = 1;
            }
        }
    }
    std::vector<long long> base_primes;
    for (long long p = 2; p <= limit; ++p) {
        if (!base_composite[p]) base_primes.push_back(p);
    }

    // Sieve the full range [2, n] in parallel, splitting into disjoint
    // segments so each thread writes to distinct array elements only.
    unsigned hc = std::thread::hardware_concurrency();
    unsigned num_threads = hc == 0 ? 4u : hc;
    num_threads = std::max(2u, std::min(num_threads, 8u));

    long long range = n - 2 + 1; // numbers from 2..n
    long long chunk = (range + num_threads - 1) / num_threads;
    if (chunk < 1) chunk = 1;

    std::vector<std::thread> threads;
    for (unsigned t = 0; t < num_threads; ++t) {
        long long lo = 2 + static_cast<long long>(t) * chunk;
        long long hi = std::min(n, lo + chunk - 1);
        if (lo > n) break;
        threads.emplace_back([lo, hi, &base_primes, &is_composite]() {
            for (long long p : base_primes) {
                long long start = ((lo + p - 1) / p) * p;
                if (start < 2 * p) start = 2 * p;
                for (long long m = start; m <= hi; m += p) {
                    is_composite[static_cast<size_t>(m)] = 1;
                }
            }
        });
    }
    for (auto& th : threads) th.join();

    // Single-threaded aggregation over the completed sieve.
    long long count = 0;
    unsigned long long sum = 0;
    long long last = -1;
    long long prev = -1;
    long long max_gap = -1;
    long long gap_prime = -1;

    for (long long i = 2; i <= n; ++i) {
        if (!is_composite[static_cast<size_t>(i)]) {
            ++count;
            sum += static_cast<unsigned long long>(i);
            last = i;
            if (prev != -1) {
                long long g = i - prev;
                if (g > max_gap) {
                    max_gap = g;
                    gap_prime = prev;
                }
            }
            prev = i;
        }
    }

    std::cout << "count " << count << "\n";
    std::cout << "sum " << sum << "\n";
    if (last == -1) {
        std::cout << "last none\n";
    } else {
        std::cout << "last " << last << "\n";
    }
    if (max_gap == -1) {
        std::cout << "gap none\n";
    } else {
        std::cout << "gap " << max_gap << " " << gap_prime << "\n";
    }

    return 0;
}

File: evidence/v1_1/ai_eval/subjects/counted/r1/sieve/rust/main.rs
use std::io::{self, Read, Write};
use std::thread;

// Reads the whole of standard input.
fn read_input() -> Vec<u8> {
    let mut bytes = Vec::new();
    io::stdin().read_to_end(&mut bytes).unwrap();
    bytes
}

fn parse_n(bytes: &[u8]) -> usize {
    let text = std::str::from_utf8(bytes).unwrap();
    text.trim().parse::<usize>().unwrap()
}

// Simple single-threaded sieve, used only to find the small primes up to
// sqrt(n) that seed the parallel sieve below.
fn small_primes_up_to(limit: usize) -> Vec<usize> {
    let mut is_composite = vec![false; limit + 1];
    if limit >= 1 {
        is_composite[0] = true;
    }
    if limit >= 1 {
        is_composite[1] = true;
    }
    let mut p = 2usize;
    while p * p <= limit {
        if !is_composite[p] {
            let mut m = p * p;
            while m <= limit {
                is_composite[m] = true;
                m += p;
            }
        }
        p += 1;
    }
    (2..=limit).filter(|&i| !is_composite[i]).collect()
}

// Sieves 2..=n, returning a bool array where index i is true when i is composite.
// The marking of composites across the range is spread over multiple threads,
// each thread owning a disjoint mutable slice of the array.
fn sieve(n: usize) -> Vec<bool> {
    let mut is_composite = vec![false; n + 1];
    if n < 2 {
        return is_composite;
    }
    is_composite[0] = true;
    is_composite[1] = true;

    let mut limit = (n as f64).sqrt() as usize;
    while (limit + 1) * (limit + 1) <= n {
        limit += 1;
    }
    while limit > 0 && limit * limit > n {
        limit -= 1;
    }
    let base_primes = small_primes_up_to(limit);

    let start_idx = 2usize;
    let len = n + 1 - start_idx;

    let available = thread::available_parallelism()
        .map(|v| v.get())
        .unwrap_or(2);
    let nthreads = available.max(2).min(8);
    let chunk_size = (len + nthreads - 1) / nthreads.max(1);
    let chunk_size = chunk_size.max(1);

    let full = &mut is_composite[start_idx..=n];
    thread::scope(|scope| {
        for (i, chunk) in full.chunks_mut(chunk_size).enumerate() {
            let offset = start_idx + i * chunk_size;
            let base_primes = &base_primes;
            scope.spawn(move || {
                let chunk_len = chunk.len();
                for &p in base_primes {
                    let mut v = if p * p > offset {
                        p * p
                    } else {
                        let rem = offset % p;
                        if rem == 0 {
                            offset
                        } else {
                            offset + (p - rem)
                        }
                    };
                    while v < offset + chunk_len {
                        chunk[v - offset] = true;
                        v += p;
                    }
                }
            });
        }
    });

    is_composite
}

fn main() {
    let input = read_input();
    let n = parse_n(&input);

    let is_composite = sieve(n);

    let mut count: u64 = 0;
    let mut sum: u64 = 0;
    let mut last: Option<usize> = None;
    let mut prev: Option<usize> = None;
    let mut max_gap: u64 = 0;
    let mut max_gap_prime: Option<usize> = None;

    if n >= 2 {
        for i in 2..=n {
            if !is_composite[i] {
                count += 1;
                sum += i as u64;
                if let Some(p) = prev {
                    let gap = (i - p) as u64;
                    if gap > max_gap {
                        max_gap = gap;
                        max_gap_prime = Some(p);
                    }
                }
                prev = Some(i);
                last = Some(i);
            }
        }
    }

    let stdout = io::stdout();
    let mut out = stdout.lock();
    writeln!(out, "count {}", count).unwrap();
    writeln!(out, "sum {}", sum).unwrap();
    match last {
        Some(l) => writeln!(out, "last {}", l).unwrap(),
        None => writeln!(out, "last none").unwrap(),
    }
    match max_gap_prime {
        Some(p) => writeln!(out, "gap {} {}", max_gap, p).unwrap(),
        None => writeln!(out, "gap none").unwrap(),
    }
}
